File: SignalKit/SignalKit/FFT.m
//
//  FFT.m
//  OScope
//
//  Created by Rob Napier on 7/17/14.
//  Copyright (c) 2014 Rob Napier. All rights reserved.
//

#import "FFT.h"
@import Accelerate;

NSArray *SpectrumForValues(NSArray *samples) {
  int Log2N = log2f(samples.count);
  int N = 1u << Log2N;
  int Stride = 1;

  double *signal = (double*)calloc(1, N * Stride * sizeof *signal);
  double *fft = (double*)calloc(1, N * sizeof *fft);

  //Fill Input Array
  for (int i=0; i < N; i++) {
    signal[i] = [samples[i] doubleValue];
  }

  UInt32 inMaxFramesPerSlice = N;
  double  mFFTNormFactor = 1.0/(2*inMaxFramesPerSlice);
  UInt32 mFFTLength = inMaxFramesPerSlice/2;
  UInt32 mLog2N = Log2N; // (int)ceilf(log2f(inMaxFramesPerSlice));

  DSPDoubleSplitComplex mDspSplitComplex;
  mDspSplitComplex.realp = (double*) calloc(mFFTLength,sizeof(double));
  mDspSplitComplex.imagp = (double*) calloc(mFFTLength, sizeof(double));
  FFTSetupD mSpectrumAnalysis = vDSP_create_fftsetupD(mLog2N, kFFTRadix2);

  double *inAudioData = signal;
  double *outFFTData = fft;

  //Generate a split complex vector from the real data
  vDSP_ctozD((DSPDoubleComplex *)inAudioData, 2, &mDspSplitComplex, 1, mFFTLength);

  //Take the fft and scale appropriately
  vDSP_fft_zripD(mSpectrumAnalysis, &mDspSplitComplex, 1, mLog2N, kFFTDirection_Forward);

  vDSP_vsmulD(mDspSplitComplex.realp, 1, &mFFTNormFactor, mDspSplitComplex.realp, 1, mFFTLength);
  vDSP_vsmulD(mDspSplitComplex.imagp, 1, &mFFTNormFactor, mDspSplitComplex.imagp, 1, mFFTLength);

  //Zero out the nyquist value
  mDspSplitComplex.imagp[0] = 0.0;

  //Convert the fft data to dB
  vDSP_zvmagsD(&mDspSplitComplex, 1, outFFTData, 1, mFFTLength);

  // FIXME: Converting to dB actually makes the graph confusing.
  //In order to avoid taking log10 of zero, an adjusting factor is added in to make the minimum value equal -128dB
  //  vDSP_vsadd(outFFTData, 1, &kAdjust0DB, outFFTData, 1, mFFTLength);
  //  Float32 one = 1;
  //  vDSP_vdbcon(outFFTData, 1, &one, outFFTData, 1, mFFTLength, 0);

  vDSP_destroy_fftsetupD(mSpectrumAnalysis);
  free (mDspSplitComplex.realp);
  free (mDspSplitComplex.imagp);

  NSMutableArray *result = [NSMutableArray array];
  for (int i =0; i < N/2; i++) {
    [result addObject:@(fft[i] * fft[i])];
  }

  free(fft);
  free(signal);

  return result;
}

File: SignalKit/SignalKit/FFT.swift
//
//  FFT.swift
//  Signal
//
//  Created by Rob Napier on 7/24/14.
//  Copyright (c) 2014 Rob Napier. All rights reserved.
//

import Accelerate

internal func spectrumForValues(signal: [Double]) -> [Double] {
  // Find the largest power of two in our samples
  let log2N = vDSP_Length(log2(Double(signal.count)))
  let n = 1 << log2N
  let fftLength = n / 2

  // This is expensive; factor it out if you need to call this function a lot
  let fftsetup = vDSP_create_fftsetupD(log2N, FFTRadix(kFFTRadix2))

  var fft = [Double](count:Int(n), repeatedValue:0.0)

  // Generate a split complex vector from the real data
  var realp = [Double](count:Int(fftLength), repeatedValue:0.0)
  var imagp = realp

  withExtendedLifetimes(realp, imagp) {
    var splitComplex = DSPDoubleSplitComplex(realp:&realp, imagp:&imagp)
    vDSP_ctozD(UnsafePointer(signal), 2, &splitComplex, 1, fftLength)

    // Take the fft
    vDSP_fft_zripD(fftsetup, &splitComplex, 1, log2N, FFTDirection(kFFTDirection_Forward))

    // Normalize
    var normFactor = 1.0 / Double(2 * n)
    vDSP_vsmulD(splitComplex.realp, 1, &normFactor, splitComplex.realp, 1, fftLength)
    vDSP_vsmulD(splitComplex.imagp, 1, &normFactor, splitComplex.imagp, 1, fftLength)

    // Zero out Nyquist
    splitComplex.imagp[0] = 0.0

    // Convert complex FFT to magnitude
    vDSP_zvmagsD(&splitComplex, 1, &fft, 1, fftLength)
  }

  // Cleanup
  vDSP_destroy_fftsetupD(fftsetup)
  return fft
}

// To get rid of the `() -> () in` casting
func withExtendedLifetime<T>(x: T, f: () -> ()) {
  return Swift.withExtendedLifetime(x, f)
}

// In the spirit of withUnsafePointers
func withExtendedLifetimes<A0, A1>(arg0: A0, arg1: A1, f: () -> ()) {
  return withExtendedLifetime(arg0) { withExtendedLifetime(arg1, f) }
}